### PR TITLE
SaltStackClient:move generating JSON to explicit method

### PR DESCRIPTION
**Pull request by hustodemon**
...to slightly reduce duplicated code in SaltStackClient methods.

### CODE REVIEW

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java [hunk 1]
```diff
@@ -97,14 +97,18 @@ public Token login(String username, String password)
...
      * @throws SaltStackException if anything goes wrong
      */
-    public Token login(String username, String password, String eauth)
+    public Token login(final String username, final String password, final String eauth)
             throws SaltStackException {
-        JsonObject json = new JsonObject();
-        json.addProperty("username", username);
-        json.addProperty("password", password);
-        json.addProperty("eauth", eauth);
+        Map<String, String> kwargs = new LinkedHashMap() {
```
Comment (moio) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: While I am a fan of this style of inline initialization of `Map`s and `List`s, I am not a big fan of `final` method parameters, [as pointed out in another PR](https://github.com/SUSE/saltstack-netapi-client-java/pull/28#commitcomment-9865167). Maybe it is just my personal taste, though. @renner, your idea?

Comment (hustodemon) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: Thanks for comment. Maybe I'm missing something, but I believe finalization of arguments doesn't leak outside of the method. It only means that arguments aren't overwriten inside the method. You can freely pass nonfinal fields to a method with final args. Isn't that correct?

Comment (renner) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: That's right @hustodemon, at least that was also my understanding of it. I would personally be fine with declaring parameters as `final` in order to ensure that these are not reassigned or changed inside that method. The code in #28 could even be simplified by that, or not? In there we explicitly create final versions of the input parameters for the async methods (`final String constUsername = username;`), which could be skipped if we declared them as final parameters.

Comment (hustodemon) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: Right, that's what I was commenting - making them final could prevent some boilerplate in #28.

The only issue with final arguments in this PR case is IMHO that it makes the method signatures even loooonger. If you think this is a problem, I'll revert back to non-final args (but will have to initialize the map the "traditional" way, i.e. without "double braces" syntax).

Comment (renner) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: I am fine with that, just wrap lines before reaching 92 characters and indent correctly. We might want to finalize the parameters of the (now merged) `*Async` methods in a separate commit or I will do it after we merged this.

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java [hunk 2]
```diff
@@ -138,32 +142,25 @@ public Token login(String username, String password, String eauth)
...
-        // Setup lowstate data to send as JSON
-        JsonObject json = new JsonObject();
-        json.addProperty("tgt", target);
-        json.addProperty("fun", function);
-        // Non-keyword arguments
-        if (args != null) {
-            JsonArray argsArray = new JsonArray();
-            for (String arg : args) {
-                argsArray.add(new JsonPrimitive(arg));
+        Map<String, String> allKwargs = new LinkedHashMap() {
```
Comment (renner) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: Please add type parameters `<String, String>` to `new LinkedHashMap()` throughout this patch.

Comment (isbm) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: Diamond operator would work too, unless we are going to support EOF'ed Java 6.

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java (hunk 1 above)
Comment (isbm) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: Personally I have no problems with the `final` keyword on the method signature and my personal taste to use it liberally. Some finds it too verbose, but others finds that the input parameters are never change, which is also good.

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java (hunk 2 above)
Comment (renner) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: @isbm: The diamond operator cannot be used with anonymous classes, see e.g. [here] (http://stackoverflow.com/questions/9773733/double-brace-initialisation-anonymous-inner-class-with-diamond-operator).

**Comment by renner** [comment]
Sorry, but could you please rebase this patch on top of the current master? I get some conflicts now when I pull the changes as #28 has been merged meanwhile. If you further fixed the missing type parameters I guess we can merge this. Thank you!

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java (hunk 2 above)
Comment (isbm) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: Bloody theory... :) Right, inference of a generic class's type arguments could influence the constraints on a generic constructor's type arguments. Well, then like you jus said: `<String, String>`.

Comment (hustodemon) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: Right, I initially wanted to use <> :)

**Comment by hustodemon** [comment]
No problem!

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java [hunk 3]
```diff
@@ -309,36 +306,25 @@ public Job call() throws SaltStackException {
...
-        // Non-keyword arguments
-        if (args != null) {
-            JsonArray argsArray = new JsonArray();
-            for (String arg : args) {
-                argsArray.add(new JsonPrimitive(arg));
+    public Map<String,Object> run(final String username, final String password,
+            final String eauth, final String client, final String target,
+            final String function, List<String> args, Map<String, String> kwargs)
+            throws SaltStackException {
+        Map<String, String> allKwags = new LinkedHashMap<String, String>() {
```
Comment (renner) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: There is a typo here, it should probably be `allKwargs`?

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java [hunk 4]
```diff
@@ -385,4 +371,5 @@ public Job call() throws SaltStackException {
         };
         return executor.submit(callable);
     }
+
```
Comment (renner) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: Please, let's not add an empty line here.

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java (hunk 3 above)
Comment (hustodemon) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: oh, no... lemme fix it.

File: src/test/java/com/suse/saltstack/netapi/utils/ClientUtilsTest.java [hunk 5]
```diff
@@ -3,6 +3,14 @@
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 
+
```
Comment (renner) [inline src/test/java/com/suse/saltstack/netapi/utils/ClientUtilsTest.java]: Please remove this line as well, let's put only one empty line to group imports.

File: src/test/java/com/suse/saltstack/netapi/utils/ClientUtilsTest.java [hunk 6]
```diff
@@ -61,4 +68,37 @@ public void testStreamToString() {
...
+        List<String> args = new ArrayList<>();
+        args.add("foo");
+        args.add("bar");
+
+        JsonObject jsonObject = ClientUtils.makeJsonData(kwargs, args);
+
+        assertEquals(expected, jsonObject);
+    }
+
```
Comment (renner) [inline src/test/java/com/suse/saltstack/netapi/utils/ClientUtilsTest.java]: ... and maybe this line as well? (sorry for nitpicking)

Comment (hustodemon) [inline src/test/java/com/suse/saltstack/netapi/utils/ClientUtilsTest.java]: no problem, will fix (I'm used to do newlines so it doesn't burn my eyes).

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java (hunk 1 above)
Comment (moio) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: Alright, I'll update my mental style guide! Thanks for discussion.

**Comment by renner** [comment]
I merged it, thanks again! Some type arguments were still missing in your test, added with 59974b1a252e596f544e25ed33c62293f2538ca8.

**Comment by hustodemon** [comment]
Oh, I totally forgot. Thanks!

**Pull request merged**